keps/sig-architecture/20191015-conformance-pr-gate.md

```diff
@@ -0,0 +1,229 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+
+We have a policy requiring Conformance Tests before promoting beta APIs to GA.
+This policy needs automation to ensure we eventually have coverage for
+everything. The proposal provides a method of measuring Conformance coverage and
+placing a `/hold` on any PR which fails to maintain or increase said coverage.
+
+
+## Motivation
+
+
+Currently PRs are merging that do not meet the Conformance Test policy
+requirement or accidentally decrease coverage by changes to how tests or
+frameworks are written.
+
+
+### Goals
+
+* Defining Conformance Coverage for API Operation 
+* Runnning Coverage per PR against kubernetes/kubernetes
+* Feedback with list of newly promoted API Operations lacking test hits
+* Feedback with list of previously tested API Operations lacking test hits
+* Soft gate with '/hold' (or similar merge blocking tag)
+* Hard gate 
+
+### Non-Goals
+
+* Defining Conformance Coverage for Watch Methods
+* Defining Conformance Coverage for Object Fields
+
+## Proposal
+
+The proposal consists of four deliverables:
+* prow-job that runs periodically/continously against master for baseline
+* prow-job that waits for PR e2e test jobs to finish for comparision
+* automated workflow for tagging / failing as a test for all PRs that reduce coverage percentage
+
+### Definition of Stable / GA Coverage
+
+APISnoop analyses Audit Logs and does not run on an active cluster yet.
+
+#### API Operations / Endpoints
+
+
+##### Total Endpoints eligible for Conformance
+
+The surface area of Kubernetes API endpoints is based on the
+paths/method/operationId of swagger.json in the k/k repo.
+
+This is not based on the generated swagger from the API server, as we analyze the audit logs statically.
+
+Out of 1100+ total, roughly 200 have been deprecated leaving 910 operations.
+
+490 of the those paths include the strings alpha and beta, which we treate as alpha and beta operations / endpoints.
+
+This currently leaves 420 'stable' endpoints.
+
+##### Test Hits
+
+The e2e.test binary adds it's user agent to calls, additionally recording the test name by appending '-- TEST CONTEXT'
+
+##### Conformance Hits
+
+When the user agent includes '--' followed by '[Conformance]' it counts as a conformance tests.
+
+#### Watch methods on Objects
+
+This may be added in the future if consensus can be reached.
+
+#### Object Fields
+
+##### PodSpec
+
+The Conformance Team has an Initial Focus will be on PodSpec. When we get closer
```

Comment:
does this mean the automation will target PodSpec versus other objects initially?

Reply:
Field coverage is still evolving.

I think the PodSpec / Field Coverage will come in a second iteration, API Operations / Endpoints will be the denominator in this first stage with ~400 endpoints.